src/pydap/responses/urs.py

```diff
@@ -0,0 +1,105 @@
+from six.moves.urllib.request import install_opener
+from six.moves.urllib.request import urlopen
+import netrc
+import re
+
+import pydap.lib
+from pydap.exceptions import ClientError
+
+import logging
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+# Set the debug level for urllib.
+debuglevel=1
+
+def install_basic_client(uri='', user='', passwd='', use_netrc=True):
+    """
+    Based on the CAS authentication example at
+    http://www.pydap.org/en/latest/client.html#authentication
+
+    Handle URS/OAuth2, and HTTP Basic authentication over https.
+    
+    To use this, you must run this function before importing the open_url()
+    function contained in pydap.client. One run, the pydap.http.request()
+    function will search a password manager for username/password pairs using
+    information passed to this function and/or from the caller's .netrc file.
+    When a HTTP request returns a authentication challenge response (401 
+    response code), this new request() function will search for credentials
+    that match the request's hostname (i.e., 'uri').
+    
+    This code knows how to follow the redirects associated with the NASA URS
+    system, an implementation of OAuth2.
+
+    :param uri: If all of uri, user and passwd are not empty, add them to the password manager.
+    :param user: See uri
+    :param passwd: See uri
+    :param netrc: If True, the default, read login credentials from the .netrc file
+    
+    """
+
+    # Create special opener with support for Cookies
+    cj = CookieJar()
+    
+    # Create the password manager and load with the credentials using 
+    pwMgr = HTTPPasswordMgrWithDefaultRealm()
+
+    # Get passwords from the .netrc file nless use_netrc is False    
+    if use_netrc:
+        logins = netrc.netrc()
+        accounts = logins.hosts # a dist of hosts and tuples
+        for host, info in accounts.items():
+            login, account, password = info
+            log.debug('Host: %s; login: %s; account: %s; password: %s' % (host, login, account, password))
+            pwMgr.add_password(None, host, login, password)
+        
+    if uri and user and passwd:
+        pwMgr.add_password(None, uri, user, passwd)
+    
+    opener = build_opener(HTTPBasicAuthHandler(pwMgr),
+                                  HTTPCookieProcessor(cj))
+    
+    opener.addheaders = [('User-agent', pydap.lib.__version__)]
+
+    install_opener(opener)
+
+    def new_request(url):
+        log.debug('Opening %s (install_basic_client)' % url)
+        r = urlopen(url)
+        
+        resp = dict([i[0], i[1]] for i in r.headers.items())
+        resp['status'] = str(r.code)
+        data = r.read()
+
+        # When an error is returned, we parse the error message from the
+        # server and return it in a ``ClientError`` exception.
+        if resp.get("content-description") == "dods_error":
+            m = re.search('code = (?P<code>\d+);\s*message = "(?P<msg>.*)"',
```

Comment:
`re` *is* imported above (line 57). But yeah, it'd be ideal to get test coverage across these lines. That's one reason to use `pydap.client.open_url()`, because you can actually pass a WSGI app as a second arg to the function. Then in the tests, you can write a fake WSGI app that just returns a dods_error, and use it to test this section of code. If you look in `src/pydap/tests/test_client.py` you can see multiple places where this strategy is employed.